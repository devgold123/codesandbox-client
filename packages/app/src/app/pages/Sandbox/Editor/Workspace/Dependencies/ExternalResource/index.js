import React from 'react';

import CrossIcon from 'react-icons/lib/md/clear';

import { EntryContainer, IconArea, Icon } from '../../elements';

const getNormalizedUrl = (url: string) => `${url.replace(/\/$/g, '')}/`;

function getName(resource: string) {
  if (resource.endsWith('.css') || resource.endsWith('.js')) {
    const match = resource.match(/.*\/(.*)/);

    if (match && match[1]) return match[1];
  }

  // Add trailing / but no double one
  return getNormalizedUrl(resource);
}

export default class ExternalResource extends React.PureComponent {
  removeResource = () => {
    this.props.removeResource(this.props.resource);
  };

  render() {
    const { resource } = this.props;
    return (
      <EntryContainer>
        <a href={resource} rel="noopener noreferrer" target="_blank">
          {getName(resource)}
        </a>
        <IconArea>
          <Icon onClick={this.removeResource}>
            <CrossIcon />
          </Icon>
        </IconArea>
      </EntryContainer>
    );
  }
}
