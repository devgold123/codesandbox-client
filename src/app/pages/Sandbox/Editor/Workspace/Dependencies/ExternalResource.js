import React from 'react';
import styled from 'styled-components';
import theme from 'common/theme';

import CrossIcon from 'react-icons/lib/md/clear';

import EntryContainer from '../EntryContainer';

const Icon = styled.div`
  transition: 0.3s ease color;
  position: absolute;
  right: 1rem;
  line-height: 1;
  color: ${() => theme.background.lighten(2).clearer(0.5)};
  &:hover {
    color: ${() => theme.red};
  }
`;

const getNormalizedUrl = (url: string) => `${url.replace(/\/$/g, '')}/`;

function getName(resource: string) {
  if (resource.endsWith('.css') || resource.endsWith('.js')) {
    return resource.match(/.*\/(.*)/)[1];
  }

  // Add trailing / but no double one
  return getNormalizedUrl(resource);
}

type Props = {
  resource: string,
  removeResource: (resource: string) => void,
};

export default class ExternalResource extends React.PureComponent {
  props: Props;

  removeResource = () => {
    this.props.removeResource(this.props.resource);
  };

  render() {
    const { resource } = this.props;
    return (
      <EntryContainer>
        <a href={resource} rel="noopener noreferrer" target="_blank">
          {getName(resource)}
        </a>
        <Icon onClick={this.removeResource}><CrossIcon /></Icon>
      </EntryContainer>
    );
  }
}
