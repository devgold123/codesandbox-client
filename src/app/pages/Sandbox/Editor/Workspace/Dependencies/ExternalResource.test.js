import React from 'react';
import renderer from 'react-test-renderer';
import ExternalResource from './ExternalResource';

describe('ExternalResource', () => {
  it('renders file if there is an extension', () => {
    const tree = renderer
      .create(<ExternalResource resource="https://ivesvh.com/ives.js" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders url if there is no extension', () => {
    const tree1 = renderer
      .create(<ExternalResource resource="https://ivesvh.com/" />)
      .toJSON();

    const tree2 = renderer
      .create(<ExternalResource resource="https://ivesvh.com" />)
      .toJSON();

    expect(tree1).toMatchSnapshot();
    expect(tree2).toMatchSnapshot();
  });
});
