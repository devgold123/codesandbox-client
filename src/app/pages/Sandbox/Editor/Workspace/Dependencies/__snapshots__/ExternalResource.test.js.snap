// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalResource renders file if there is an extension 1`] = `
<span
  className="sc-bdVaJa jtVJZx"
>
  <a
    href="https://ivesvh.com/ives.js"
    rel="noopener noreferrer"
    target="_blank"
  >
    ives.js
  </a>
  <div
    className="sc-bwzfXH gaVNMC"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 40 40"
      width="1em"
    >
      <g>
        <path
          d="m31.6 10.7l-9.3 9.3 9.3 9.3-2.3 2.3-9.3-9.3-9.3 9.3-2.3-2.3 9.3-9.3-9.3-9.3 2.3-2.3 9.3 9.3 9.3-9.3z"
        />
      </g>
    </svg>
  </div>
</span>
`;

exports[`ExternalResource renders url if there is no extension 1`] = `
<span
  className="sc-bdVaJa jtVJZx"
>
  <a
    href="https://ivesvh.com/"
    rel="noopener noreferrer"
    target="_blank"
  >
    https://ivesvh.com/
  </a>
  <div
    className="sc-bwzfXH gaVNMC"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 40 40"
      width="1em"
    >
      <g>
        <path
          d="m31.6 10.7l-9.3 9.3 9.3 9.3-2.3 2.3-9.3-9.3-9.3 9.3-2.3-2.3 9.3-9.3-9.3-9.3 2.3-2.3 9.3 9.3 9.3-9.3z"
        />
      </g>
    </svg>
  </div>
</span>
`;

exports[`ExternalResource renders url if there is no extension 2`] = `
<span
  className="sc-bdVaJa jtVJZx"
>
  <a
    href="https://ivesvh.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    https://ivesvh.com/
  </a>
  <div
    className="sc-bwzfXH gaVNMC"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 40 40"
      width="1em"
    >
      <g>
        <path
          d="m31.6 10.7l-9.3 9.3 9.3 9.3-2.3 2.3-9.3-9.3-9.3 9.3-2.3-2.3 9.3-9.3-9.3-9.3 2.3-2.3 9.3 9.3 9.3-9.3z"
        />
      </g>
    </svg>
  </div>
</span>
`;
